### PR TITLE
Provide a way to set a response timeout and attributes using `WebClientRequestPreparation`

### LINKED ISSUE #3347: Provide a way to configure a response timeout when building an `HttpRequest`
If a user needs to dynamically change a response timeout, he/she has to use a context customizer.
For example:
```java
try (SafeCloseable ignored = Clients.withContextCustomizer(ctx -> {
    ctx.setResponseTimeoutMillis(timeoutMillis);
})) {
    client.execute(req)...
}
```

The API mutates the response timeout of `ClientRequestContext` before executing. But the API is not user-friendly.
It would be nice if users can set a response timeout in a declarative way with a `HttpRequestBuilder`
```java
HttpRequest.builder()
           .responseTimeoutMillis(...)
```

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

If a response timeout is set with a `HttpRequestBuilder`, it would be convenient when a single `WebClient` is used.
Furthermore, a user who only uses Armeria WebClient without Armeria Server can easily set `Attribute`s without `Clients.withContextCustomizer()`.

Modifications:

- Add `responseTimeout{Millis}()` to `WebClientRequestPreparation` for fluently setting a response timeout
- Add `setAttr()` to `WebClientRequestPreparation` for fluently setting attributes
- Configure a response timeout and attributes using `Clients.withContextCustomizer()` before executing a request.

Result:

- You can now set a response timeout and attributes using `WebClientRequestPreparation`
  ```java
  WebClient client = ...;
  client.prepare()
        .get("/my-service")
        .responseTimeout(Duration.ofSeconds(3))
        .setAttr(USER_ID, userId)
        .setAttr(USER_SECRET, secret)
        .execute();
  ```

- Fixes #3347

### CODE REVIEW

**Comment by minwoox** [comment]
Should the `RequestOptions` have all options that a user can set via `ClientRequestContext`?
(e.g. writeTimeout, maxResponseLength, additional headers)

**Comment by ikhoon** [comment]
> Should the `RequestOptions` have all options that a user can set via `ClientRequestContext`?
> (e.g. writeTimeout, maxResponseLength, additional headers)

I thought they are not commonly configured at the request level. Any thoughts? @imasahiro @jrhee17

**Comment by trustin** [comment]
> Add `RequestOptions options()` to `HttpRequest` for representing the current request options

Wouldn't this open a can of worms for us? `HttpRequest` is a common interface that can be used for both client- and server-side as well as the case where no networking takes place. I think it's better keeping `HttpRequest` independent from any networking-level behaviors.

How about this?

- Add the convenience methods to `WebClientRequestPreparation` only.
- Don't store the new properties in `HttpRequest`. Keep them only in `WebClientRequestPreparation` and use them when a user calls `execute()`.

**Comment by ikhoon** [comment]
> Wouldn't this open a can of worms for us? HttpRequest is a common interface that can be used for both client- and server-side as well as the case where no networking takes place.

That was one of my considering points. It sounds good to remove `RequestOptions` and use the configuration only in `WebClientRequestPreparation`

File: core/src/main/java/com/linecorp/armeria/client/WebClientRequestPreparation.java [hunk 1]
```diff
@@ -46,7 +64,76 @@
...
+        this.responseTimeoutMillis = responseTimeoutMillis;
+        return this;
+    }
+
+    /**
+     * Associates the specified value with the given {@link AttributeKey} in this request.
+     * If this context previously contained a mapping for the {@link AttributeKey}, the old value is replaced
+     * by the specified value.
+     */
+    public <V> WebClientRequestPreparation setAttr(AttributeKey<V> key, @Nullable V value) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/WebClientRequestPreparation.java:124]: How about just `attr()`?

File: core/src/main/java/com/linecorp/armeria/client/WebClientRequestPreparation.java [hunk 2]
```diff
@@ -46,7 +64,76 @@
...
+                return client.execute(httpRequest);
+            }
+        } else {
+            return client.execute(httpRequest);
+        }
+    }
+
+    @Nullable
+    private Consumer<ClientRequestContext> newContextCustomizer() {
+        if (responseTimeoutMillis == -1 && attributes == null) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClientRequestPreparation.java:81]:
```suggestion
        if (responseTimeoutMillis == -1 && (attributes == null || attributes.isEmpty())) {
```

**Review by minwoox — APPROVED** [review]
Just one nit. Thanks! 😄

**Review by trustin — APPROVED** [review]
Thanks, @ikhoon :bow: Nice usability boost!

**Pull request merged**